source/common/router/scoped_config_impl.cc

```diff
@@ -16,65 +18,112 @@ bool ScopeKey::operator==(const ScopeKey& other) const {
+}
+
+std::unique_ptr<ScopeKeyFragmentBase> FragmentBuilderImpl::computeFragmentFromHeader(
+    const Http::HeaderMap& headers,
+    const ScopedRoutes::ScopeKeyBuilder::FragmentBuilder::HeaderValueExtractor&
+        header_value_extractor_config) const {
+
+  const auto& header_entry =
+      headers.get(Envoy::Http::LowerCaseString(header_value_extractor_config.name()));
   if (header_entry.empty()) {
     return nullptr;
   }
 
   // This is an implicitly untrusted header, so per the API documentation only the first
   // value is used.
   std::vector<absl::string_view> elements{header_entry[0]->value().getStringView()};
-  if (header_value_extractor_config_.element_separator().length() > 0) {
+  if (header_value_extractor_config.element_separator().length() > 0) {
     elements = absl::StrSplit(header_entry[0]->value().getStringView(),
-                              header_value_extractor_config_.element_separator());
+                              header_value_extractor_config.element_separator());
   }
-  switch (header_value_extractor_config_.extract_type_case()) {
+  switch (header_value_extractor_config.extract_type_case()) {
   case ScopedRoutes::ScopeKeyBuilder::FragmentBuilder::HeaderValueExtractor::kElement:
     for (const auto& element : elements) {
       std::pair<absl::string_view, absl::string_view> key_value = absl::StrSplit(
-          element, absl::MaxSplits(header_value_extractor_config_.element().separator(), 1));
-      if (key_value.first == header_value_extractor_config_.element().key()) {
+          element, absl::MaxSplits(header_value_extractor_config.element().separator(), 1));
+      if (key_value.first == header_value_extractor_config.element().key()) {
         return std::make_unique<StringKeyFragment>(key_value.second);
       }
     }
     break;
   case ScopedRoutes::ScopeKeyBuilder::FragmentBuilder::HeaderValueExtractor::kIndex:
-    if (header_value_extractor_config_.index() < elements.size()) {
-      return std::make_unique<StringKeyFragment>(elements[header_value_extractor_config_.index()]);
+    if (header_value_extractor_config.index() < elements.size()) {
+      return std::make_unique<StringKeyFragment>(elements[header_value_extractor_config.index()]);
     }
     break;
-  default:                       // EXTRACT_TYPE_NOT_SET
-    NOT_REACHED_GCOVR_EXCL_LINE; // Caught in constructor already.
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
   }
 
   return nullptr;
 }
 
+std::unique_ptr<ScopeKeyFragmentBase>
+FragmentBuilderImpl::computeFragmentFromMetadata(const Metadata& meta,
+                                                 const MetadataValueExtractorConfig& config) const {
+  const auto& value = Envoy::Config::Metadata::metadataValue(&meta, config.metadata_key());
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::kStringValue:
+    return std::make_unique<StringKeyFragment>(value.string_value());
+  default:
+    // TODO (tonya11en): Support structured metadata fragments.
```

Comment:
Sorry what does this mean? Do we need to raise this as an error/warning/etc.? Or at least a debug log?

Reply:
Folks can potentially pack [typed metadata](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/data_sharing_between_filters.html#typed-metadata) in the map (like kList or kStruct types) as opposed to just string values. There's potential here for more complicated matching logic based on these more complex/custom types.